shared_model/validators/transaction_batch_validator.cpp

```diff
@@ -46,19 +49,44 @@ namespace {
       return BatchCheckResult::kNoBatchMeta;
     }
 
+    bool batch_is_ordered = batch_meta_opt->get()->type()
+        == shared_model::interface::types::BatchType::ORDERED;
+
     const auto &batch_hashes = batch_meta_opt->get()->reducedHashes();
-    if (batch_hashes.size() != transactions_quantity) {
-      return BatchCheckResult::kIncorrectBatchMetaSize;
-    }
+    bool hashes_are_correct = true;
+
+    auto hashes_begin = boost::begin(batch_hashes);
+    auto hashes_end = boost::end(batch_hashes);
+    auto transactions_begin = boost::begin(transactions);
+    auto transactions_end = boost::end(transactions);
+
+    if (batch_is_ordered and partial_ordered_batches_are_valid) {
+      // checking that transactions' hashes are subsequence of batch meta hashes
+      for (; transactions_begin != transactions_end; ++hashes_begin) {
+        if (hashes_begin == hashes_end) {
+          hashes_are_correct = false;
+          break;
+        }
+        if (*hashes_begin == transactions_begin->reducedHash()) {
+          ++transactions_begin;
+        }
+      }
+    } else {
```

Comment:
Seems this `if` statement should be separated on different validators. For now, is ok to leave a todo about reworking.